src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -34,7 +34,31 @@ public async Task RunAsync(ActionRunStage stage)
             Trace.Entering();
             ArgUtil.NotNull(ExecutionContext, nameof(ExecutionContext));
             ArgUtil.NotNull(Inputs, nameof(Inputs));
-            ArgUtil.NotNull(Data.Steps, nameof(Data.Steps));
+
+            List<Pipelines.ActionStep> steps;
+            if (stage == ActionRunStage.Pre)
+            {
+                ArgUtil.NotNull(Data.PreSteps, nameof(Data.PreSteps));
+                steps = Data.PreSteps;
+            } 
+            else if (stage == ActionRunStage.Post)
+            {
+                ArgUtil.NotNull(Data.PostSteps, nameof(Data.PostSteps));
+                steps = Data.PostSteps.ToList();
+                // Only register post steps for steps that actually ran
+                for (int i = 0; i < steps.Count; i++)
+                {
+                    if (!ExecutionContext.Root.ChildStepsWithPostRegistered.Contains(steps[i].Id))
+                    {
+                        steps.RemoveAt(i);
```

Comment:
hard to tell, but this might cause index out of bounds (or skip an element). consider `.Where(x => ...)`